ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,252 @@
+      throw new KsqlException(
+          "Expected a value for each column. Expected Columns: " + columns
+              + ". Got " + insertValues.getValues());
+    }
+
+    final Map<String, Object> values = new HashMap<>();
+    for (int i = 0; i < columns.size(); i++) {
+      final String column = columns.get(i);
+      final Schema columnSchema = dataSource.getSchema().field(column).schema();
+      final Expression valueExp = insertValues.getValues().get(i);
+
+      values.put(column, new ExpressionResolver(columnSchema, column).process(valueExp, null));
+    }
+
+    if (keyField.isPresent()) {
+      final String key = keyField.get();
+      final Object keyValue = values.get(key);
+      final Object rowKeyValue = values.get(SchemaUtil.ROWKEY_NAME);
+
+      if ((keyValue != null) ^ (rowKeyValue != null)) {
+        values.putIfAbsent(key, rowKeyValue);
+        values.putIfAbsent(SchemaUtil.ROWKEY_NAME, keyValue);
+      } else if (!Objects.equals(keyValue, rowKeyValue)) {
+        throw new KsqlException(
+            String.format(
+                "Expected ROWKEY and %s to match but got %s and %s respectively.",
+                key, rowKeyValue, keyValue));
+      }
+    }
+
+    values.putIfAbsent(SchemaUtil.ROWTIME_NAME, clock.getAsLong());
+
+    for (Field field : dataSource.getSchema().fields()) {
+      if (!field.schema().isOptional() && values.getOrDefault(field.name(), null) == null) {
+        throw new KsqlException("Got null value for nonnull field: " + field);
+      }
+    }
+
+    return new GenericRow(
+        dataSource.getSchema()
+            .fields()
+            .stream()
+            .map(Field::name)
+            .map(values::get)
+            .collect(Collectors.toList())
+    );
+  }
+
+  @SuppressWarnings("unchecked") // we know that key is String
+  private byte[] serializeKey(final Object keyValue, final StructuredDataSource<?> dataSource) {
+    if (keyValue == null) {
+      return null;
+    }
+
+    return ((Serde<String>) dataSource.getKeySerdeFactory().create())
+        .serializer()
+        .serialize(dataSource.getKafkaTopicName(), keyValue.toString());
+  }
+
+  private byte[] serializeRow(
+      final GenericRow row,
+      final StructuredDataSource<?> dataSource,
+      final KsqlConfig config,
+      final ServiceContext serviceContext
+  ) {
+    final Serde<GenericRow> rowSerde = dataSource.getKsqlTopicSerde()
+        .getGenericRowSerde(
+            dataSource.getSchema(),
+            config,
+            serviceContext.getSchemaRegistryClientFactory(),
+            "",
+            NoopProcessingLogContext.INSTANCE);
+    return rowSerde.serializer().serialize(dataSource.getKafkaTopicName(), row);
+  }
+
+  private static class ExpressionResolver extends AstVisitor<Object, Void> {
+
+    private static final Map<Type, Function<Number, Object>> CASTER =
+        ImmutableMap.<Type, Function<Number, Object>>builder()
+            .put(Type.INT8, Number::shortValue)
```

Comment:
KSQL doesn't support INT8... right?

Reply:
I just removed up-casting entirely. I can add that in a different PR.